plugins.crunchyroll: update/fix

mkbloke opened the pull request:
* update params for Windows app
* remove Android headers
* fix device_id cached value
* add id to metadata
* minor spelling corrections
* add support for beta URLs
* update tests
* update cache.set() calls to use expires kw
* update expires_at=0 to expires=0

closes #3674
closes #4453

bastimeyer reviewed (commented):
Since you're changing the `_access_type` from the Android app to the Windows desktop app, it's probably not a bad idea to change the `_user_agent` (or alternatively, import from the existing user agents)

mkbloke commented:
Initially, I thought what you wrote about updating the user agent was a good idea, but there are also a lot of other Android related headers sent at the moment too.  The thing is, I can't test this plugin properly myself as I don't have an account with Crunchyroll to do so.  https://github.com/streamlink/streamlink/issues/4453#issuecomment-1114520315 suggests that these changes alone result in a working plugin, so I'm reluctant to make changes without being able to test fully.

I could look at the Windows app traffic and copy the UA and headers from it so that the plugin looks and acts like the app, although I'll need to get folks in #4453 to help test it fully.

mkbloke commented:
Some things to note:

`start_session` is a GET request in the Windows app rather than a POST, but POST works fine, so I have left it like that for the time being.

The HTTP client in the Windows app does not send many headers.  It doesn't sent `Host:`, even though it's using HTTPS.  It doesn't send `User-Agent:` either.  I tried not sending the UA by popping it from `self.session.http.headers`, but that just reverts to the Python requests default UA string.  Setting it to a zero length string causes `User-Agent:` to be sent with no value.  I have left it to send the default Firefox UA for the time being, which is probably fine.

It seems like there has been a minor bug since the refactor in #1820 which caused the `device_id` value to be set to the expires value.  Strange how it's gone unnoticed for so long, but obviously not of any great consequence.

I've tested this on a free stream, but cannot test the login and therefore access to premium content.  I will ask if folks can test it fully in #4453.  Hopefully that will all still work correctly, as little has changed and the request sending the login data is the same as that used by the Windows app.

mkbloke commented:
I've added the `WIP` label again... bear with me.

I've added some code to also try to close #3674, which will need testing with a selection of beta URLs, as I only have access to those in the issue.  I've also updated the tests to use `should_match_groups[]` and the added the beta URLs in the tests.

Hopefully not too much more to do.

back-to commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -148,17 +135,18 @@ def _api_call(self, entrypoint, params=None, schema=None):
 
         # The certificate used by Crunchyroll cannot be verified in some environments.
-        res = self.session.http.post(url, data=params, headers=self.headers, verify=False)
+        res = self.session.http.post(url, data=params, verify=False)
```
could you remove the `verify=False` and the comment above

bastimeyer commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -305,9 +300,33 @@ def stream_weight(cls, key):
+                    )),
+                )),
+            ))
+            if beta_id not in json:
+                return
```
If `json` is `None` (which you're explicitly validating), then this will raise `TypeError: argument of type 'NoneType' is not iterable`.

bastimeyer commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -248,10 +236,17 @@ def get_info(self, media_id, fields=None, schema=None):
+    )
 """, re.VERBOSE))
 class Crunchyroll(Plugin):
+    beta_json_re = re.compile(r"window.__INITIAL_STATE__\s*=\s*({.*});")
```
Don't put this regex here. Just define it in `_get_streams`

bastimeyer commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -305,9 +300,33 @@ def stream_weight(cls, key):
+                        validate.any(None, validate.all(
+                            {"content": {"byId": {str: {"external_id": validate.all(
+                                validate.transform(lambda s: int(s.replace("EPI.", ""))),
+                                int,
```
The `int` type validation is unnecessary, as you're already transforming to `int` above

bastimeyer commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -305,9 +300,33 @@ def stream_weight(cls, key):
+            json = self.session.http.get(self.url, schema=validate.Schema(
+                validate.parse_html(),
+                validate.xml_xpath_string(".//script[contains(text(), 'window.__INITIAL_STATE__')]/text()"),
+                validate.any(None, validate.all(
+                    validate.transform(self.beta_json_re.search),
+                    validate.any(None, validate.all(
```
This could get simplified by catching `PluginError`s from the validation and validating the correct types instead of `validate.any(None, ...)`, namely `str` and `re.Match`. But ignore this for now.

bastimeyer reviewed (commented):
While your changes will fix the plugin issue according to the user responses, there's still a ton of things wrong with the current plugin implementation, because it hasn't been properly maintained and updated in years. The plugin should get rewritten from scratch at some point in the future, to fix all the other issues. Fixing the current issues should be good for now though.

mkbloke commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -352,14 +372,12 @@ def _get_streams(self):
         return streams
 
     def _create_api(self):
-        """Creates a new CrunchyrollAPI object, initiates it's session and
+        """Creates a new CrunchyrollAPI object, initiates its session and
         tries to authenticate it either by using saved credentials or the
         user's username and password.
         """
         if self.options.get("purge_credentials"):
-            self.cache.set("session_id", None, 0)
-            self.cache.set("auth", None, 0)
-            self.cache.set("session_id", None, 0)
+            self.cache.set("auth", None, expires=0)
```
I'm not sure why `self.cache.set("session_id", None, 0)` was called twice here.  I'm not sure why it's called at all, as there is nothing that sets the `session_id` in the cache.

I wonder if it is worth also scrubbing the `device_id` from the cache when the credentials are purged?

bastimeyer replied:
> was called twice here

https://github.com/streamlink/streamlink/blame/bcd624c1753404e7119837a4e050f785485818d3/src/streamlink/plugins/crunchyroll.py#L360-L362
Probably a mistake in 1ced3256ae257ae7bb19675a6d68ac24aa4e7a6c

> also scrubbing the `device_id` from the cache

Purge credentials should clear the entire cache.

bastimeyer commented on src/streamlink/plugins/crunchyroll.py:
```diff
@@ -305,9 +299,34 @@ def stream_weight(cls, key):
+                    )),
+                )),
+            ))
+            if not json:
+                return
+            media_id = json[beta_id]["external_id"]
```
```suggestion
            if not json or beta_id not in json:
                return
            media_id = json[beta_id]["external_id"]
```

bastimeyer commented:
@mkbloke Do you want to wait for some responses to [your question in regards to the site's beta version](https://github.com/streamlink/streamlink/issues/3674#issuecomment-1120010035), or do you think that the current changes are good enough and can get merged into master?

mkbloke commented:
A couple of people have reported feedback on the beta URL support in issue #4453 so far to say that it's working for them, so, I think the code is *probably* OK to be merged as is, but I'm also happy to wait to a bit longer for more testing and possible feedback on it.
